File: packages/pvm-assemblyscript/assembly/instructions/control-flow.ts
/**
 * Control Flow Instructions
 *
 * NOP, HALT, ERROR, CALL, RETURN, JUMP, JUMP_IF, JUMP_IF_NOT
 */

import {
  HALT_ADDRESS,
  OPCODE_FALLTHROUGH,
  OPCODE_JUMP,
  OPCODE_JUMP_IND,
  OPCODE_LOAD_IMM_JUMP,
  OPCODE_LOAD_IMM_JUMP_IND,
  OPCODE_TRAP,
  RESULT_CODE_HALT,
  RESULT_CODE_PANIC,
} from '../config'
import { InstructionContext, InstructionResult } from '../types'
import { BaseInstruction } from './base'

/**
 * TRAP instruction (opcode 0x00)
 * Panics the PVM as specified in Gray Paper
 */
export class TRAPInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_TRAP
  name: string = 'TRAP'

  execute(_context: InstructionContext): InstructionResult {
    return new InstructionResult(RESULT_CODE_PANIC)
  }
}

/**
 * FALLTHROUGH instruction (opcode 0x01)
 * No operation as specified in Gray Paper
 */
export class FALLTHROUGHInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_FALLTHROUGH
  name: string = 'FALLTHROUGH'

  execute(_context: InstructionContext): InstructionResult {
    return new InstructionResult(-1)
  }
}

/**
 * JUMP instruction (opcode 0x40)
 * Unconditional jump with offset as specified in Gray Paper
 * 
 * Gray Paper (pvm.tex line 322):
 * Format: "One Offset" (lines 308-314)
 * immed_X ≡ ι + signfunc{l_X}(decode[l_X]{instructions[ι+1:l_X]})
 * Mutation: branch(immed_X, ⊤)
 */
export class JUMPInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_JUMP
  name: string = 'JUMP'

  execute(context: InstructionContext): InstructionResult {
    // Gray Paper: Instructions with Arguments of One Offset
    // immed_X = ι + signfunc{l_X}(decode[l_X]{instructions[ι+1:l_X]})
    const targetAddress = this.parseOneOffset(
      context.operands,
      context.fskip,
      context.programCounter,
    )

    // Gray Paper: Static jumps must target basic block starts
    // Use validateBranchTarget for consistent validation with other branch instructions
    const validationResult = this.validateBranchTarget(targetAddress, context)
    if (validationResult !== null) {
      return validationResult
    }

    // Mutate context directly
    context.programCounter = targetAddress

    return new InstructionResult(-1)
  }
}

/**
 * JUMP_IND instruction (opcode 0x50)
 * Indirect jump as specified in Gray Paper
 *
 * Gray Paper (pvm.tex line 343):
 * Format: "One Register & One Immediate" (lines 326-335)
 * r_A = min(12, instructions[ι+1] mod 16)
 * l_X = min(4, max(0, ℓ - 1))
 * immed_X ≡ sext{l_X}{decode[l_X]{instructions[ι+2:l_X]}}
 * Mutation: djump((reg_A + immed_X) mod 2^32)
 *
 * Where djump(a) is defined in equations 211-216:
 * djump(a) ⟹ (ε, ι') = {
 *   (halt, ι)                                 when a = 2^32 - 2^16
 *   (panic, ι)                                when a = 0 ∨ a > len(j)·2 ∨ a mod 2 ≠ 0 ∨ j_{(a/2)-1} ∉ basicblocks
 *   (continue, j_{(a/2)-1})                   otherwise
 * }
 */
export class JUMP_INDInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_JUMP_IND
  name: string = 'JUMP_IND'

  execute(context: InstructionContext): InstructionResult {
    const parseResult = this.parseOneRegisterAndImmediate(context.operands, context.fskip)
    const registerA: u8 = parseResult.registerA
    const immediateX = parseResult.immediateX
    const registerValue = this.getRegisterValueAs64(
      context.registers,
      registerA,
    )

    // Gray Paper djump logic (equation 211-216):
    // djump(a) ⟹ (ε, ι') = {
    //   (halt, ι) when a = 2^32 - 2^16
    //   (panic, ι) otherwhen a = 0 ∨ a > len(j)·2 ∨ a mod 2 ≠ 0 ∨ j_{(a/2) - 1} ∉ basicblocks
    //   (continue, j_{(a/2) - 1}) otherwise
    // }
    // a = (register + immediateX) % 2^32
    const a = u64((registerValue + immediateX) & u64(0xffffffff))

    // Check for HALT condition: a = 2^32 - 2^16
    if (a === u64(HALT_ADDRESS)) {
      return new InstructionResult(RESULT_CODE_HALT)
    }

    // Check for PANIC conditions:
    // - a = 0
    // - a > len(j) × 2
    // - a mod 2 ≠ 0
    const maxAddress = u32(context.jumpTable.length) * 2
    if (a === u64(0) || a > u64(maxAddress) || (a % u64(2)) !== u64(0)) {
      return new InstructionResult(RESULT_CODE_PANIC)
    }

    // Calculate jump table index: (a / 2) - 1
    const index = i32((a / u64(2)) - u64(1))

    // Check if index is valid
    if (index < 0 || index >= context.jumpTable.length) {
      return new InstructionResult(RESULT_CODE_PANIC)
    }

    // Get target from jump table
    const targetAddress = context.jumpTable[index]

    // Gray Paper: Validate that target is a valid basic block start
    // This is the same validation as LOAD_IMM_JUMP and other branch instructions
    // basicblocks = ({0} ∪ {n + 1 + Fskip(n) : n is termination instruction})
    //               ∩ {n : bitmask[n] = 1 and c[n] in valid opcodes}
    const validationResult = this.validateBranchTarget(targetAddress, context)
    if (validationResult !== null) {
      return validationResult
    }

    // Mutate context directly
    context.programCounter = targetAddress

    return new InstructionResult(-1)
  }
}

/**
 * LOAD_IMM_JUMP instruction (opcode 0x80)
 * Load immediate and jump as specified in Gray Paper
 * 
 * Gray Paper (pvm.tex line 404):
 * Format: "One Register, One Immediate and One Offset" (lines 385-396)
 * r_A = min(12, instructions[ι+1] mod 16)
 * l_X = min(4, ⌊instructions[ι+1]/16⌋ mod 8)
 * immed_X = sext{l_X}{decode[l_X]{instructions[ι+2:l_X]}}
 * l_Y = min(4, max(0, ℓ - l_X - 1))
 * immed_Y = ι + signfunc{l_Y}(decode[l_Y]{instructions[ι+2+l_X:l_Y]})
 * Mutation: branch(immed_Y, ⊤), reg_A' = immed_X
 */
export class LOAD_IMM_JUMPInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_LOAD_IMM_JUMP
  name: string = 'LOAD_IMM_JUMP'

  execute(context: InstructionContext): InstructionResult {
    const parseResult = this.parseRegisterAndTwoImmediates(
        context.operands,
        context.fskip,
      )
    const registerA: u8 = parseResult.registerA
    const immediateX = parseResult.immediateX
    const immediateY = parseResult.immediateY

    // Interpret immediateY as a signed offset
    // If MSB is set in the original value, treat it as negative
    // This ensures context.programCounter + signedOffset effectively subtracts when needed
    const signedOffset = this.toSigned64(immediateY)
    const targetAddress = u32(i32(context.programCounter) + i32(signedOffset))

    // Gray Paper: Static jumps must target basic block starts
    // Use validateBranchTarget to check if target is in basicblocks
    // Gray Paper equation 200-204: branch(b, C) → panic if b not in basicblocks
    // basicblocks = ({0} ∪ {n + 1 + Fskip(n) : n is termination instruction})
    //               ∩ {n : bitmask[n] = 1 and c[n] in valid opcodes}
    const validationResult = this.validateBranchTarget(targetAddress, context)
    if (validationResult !== null) {
      return validationResult
    }

    // Set register A to immediateX (Gray Paper: reg_A' = immed_X)
    this.setRegisterValueWith64BitResult(
      context.registers,
      registerA,
      immediateX,
    )

    // Mutate context directly - set PC to target address
    context.programCounter = targetAddress

    return new InstructionResult(-1)
  }
}

/**
 * LOAD_IMM_JUMP_IND instruction (opcode 0x180)
 * Load immediate and indirect jump as specified in Gray Paper
 * 
 * Gray Paper (pvm.tex lines 583-586):
 * Format: "Two Registers and Two Immediates" (lines 560-575)
 * r_A = min(12, instructions[ι+1] mod 16)
 * r_B = min(12, ⌊instructions[ι+1]/16⌋)
 * l_X = min(4, instructions[ι+2] mod 8)
 * immed_X = sext{l_X}{decode[l_X]{instructions[ι+3:l_X]}}
 * l_Y = min(4, max(0, ℓ - l_X - 2))
 * immed_Y = sext{l_Y}{decode[l_Y]{instructions[ι+3+l_X:l_Y]}}
 * Mutation: djump((reg_B + immed_Y) mod 2^32), reg_A' = immed_X
 */
export class LOAD_IMM_JUMP_INDInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_LOAD_IMM_JUMP_IND
  name: string = 'LOAD_IMM_JUMP_IND'

  execute(context: InstructionContext): InstructionResult {
    // Gray Paper: Instructions with Arguments of Two Registers and Two Immediates
    // r_A = min(12, (instructions[ι+1]) mod 16)
    // r_B = min(12, ⌊instructions[ι+1]/16⌋)
    // l_X = min(4, instructions[ι+2] mod 8)
    // immed_X = sext{l_X}{decode[l_X]{instructions[ι+3:l_X]}}
    // l_Y = min(4, max(0, ℓ - l_X - 2))
    // immed_Y = sext{l_Y}{decode[l_Y]{instructions[ι+3+l_X:l_Y]}}
    const parseResult = this.parseTwoRegistersAndTwoImmediates(
        context.operands,
        context.fskip,
      )
    const registerA = parseResult.registerA
    const registerB = parseResult.registerB
    const immediateX = parseResult.immediateX
    const immediateY = parseResult.immediateY

    // Gray Paper: djump((reg_B + immed_Y) mod 2^32)
    // Read register B value BEFORE potentially overwriting it
    const registerBValue = this.getRegisterValueAs64(
      context.registers,
      registerB,
    )

    // Gray Paper: reg_A' = immed_X
    this.setRegisterValueWith64BitResult(
      context.registers,
      registerA,
      immediateX,
    )

    const a = u64((registerBValue + immediateY) & u64(0xffffffff)) // mod 2^32

    // Gray Paper djump logic (equation 11):
    // Check for HALT condition: a = 2^32 - 2^16
    if (a === u64(HALT_ADDRESS)) {
      return new InstructionResult(RESULT_CODE_HALT)
    }

    // Check for PANIC conditions:
    // - a = 0
    // - a > len(j) × 2
    // - a mod 2 ≠ 0
    const maxAddress = u32(context.jumpTable.length) * 2
    if (a === u64(0) || a > u64(maxAddress) || (a % u64(2)) !== u64(0)) {
      return new InstructionResult(RESULT_CODE_PANIC)
    }

    // Calculate jump table index: (a / 2) - 1
    const index = i32((a / u64(2)) - u64(1))

    // Check if index is valid
    if (index < 0 || index >= context.jumpTable.length) {
      return new InstructionResult(RESULT_CODE_PANIC)
    }

    // Get target from jump table
    const targetAddress = context.jumpTable[index]

    // Gray Paper: Validate that target is a valid basic block start
    // This is the same validation as LOAD_IMM_JUMP, JUMP_IND, and other branch instructions
    // According to Gray Paper equation 11: j[(a/2)-1] ∉ basicblocks → panic
    // basicblocks = ({0} ∪ {n + 1 + Fskip(n) : n is termination instruction})
    //               ∩ {n : bitmask[n] = 1 and c[n] in valid opcodes}
    const validationResult = this.validateBranchTarget(targetAddress, context)
    if (validationResult !== null) {
      return validationResult
    }

    context.programCounter = targetAddress

    return new InstructionResult(-1)
  }
}

File: packages/pvm-rust/src/instructions/control_flow.rs
//! Control flow instructions (TRAP, FALLTHROUGH, JUMP, etc.). Mirrors assembly/instructions/control-flow.ts.

use crate::config::{
    HALT_ADDRESS, OPCODE_FALLTHROUGH, OPCODE_JUMP, OPCODE_JUMP_IND, OPCODE_LOAD_IMM_JUMP,
    OPCODE_LOAD_IMM_JUMP_IND, OPCODE_TRAP, RESULT_CODE_HALT, RESULT_CODE_PANIC,
};
use crate::instructions::base::{
    parse_one_offset, parse_one_register_and_immediate, parse_register_and_two_immediates,
    parse_two_registers_and_two_immediates, validate_branch_target, InstructionHandler,
};
use crate::types::{InstructionContext, InstructionResult};

/// TRAP (opcode 0): panic.
pub struct TrapInstruction;

impl TrapInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for TrapInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_TRAP)
    }

    fn name(&self) -> &'static str {
        "TRAP"
    }

    fn execute(&self, _context: &mut InstructionContext<'_>) -> InstructionResult {
        InstructionResult::new(RESULT_CODE_PANIC as i32, 0)
    }
}

impl Default for TrapInstruction {
    fn default() -> Self {
        Self::new()
    }
}

/// FALLTHROUGH (opcode 1): no-op, continue.
pub struct FallthroughInstruction;

impl FallthroughInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for FallthroughInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_FALLTHROUGH)
    }

    fn name(&self) -> &'static str {
        "FALLTHROUGH"
    }

    fn execute(&self, _context: &mut InstructionContext<'_>) -> InstructionResult {
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

impl Default for FallthroughInstruction {
    fn default() -> Self {
        Self::new()
    }
}

/// JUMP (opcode 0x40): unconditional jump. Gray Paper one-offset format.
/// Format: One Offset. immed_X = ι + signfunc(offset). Mutation: branch(immed_X, ⊤).
/// Checks: target must be a valid basic block start (validate_branch_target).
pub struct JumpInstruction;

impl JumpInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for JumpInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_JUMP)
    }

    fn name(&self) -> &'static str {
        "JUMP"
    }

    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        // Gray Paper: target = ι + signfunc(offset); use instruction PC (current program_counter).
        let target = parse_one_offset(context.operands, context.fskip, context.program_counter);
        if let Some(panic_result) =
            validate_branch_target(target, context.code, context.bitmask)
        {
            return panic_result;
        }
        context.program_counter = target;
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

impl Default for JumpInstruction {
    fn default() -> Self {
        Self::new()
    }
}

/// JUMP_IND (opcode 0x50): indirect jump. djump((reg_A + immed_X) mod 2^32).
/// Gray Paper djump(a): HALT when a = 2^32-2^16; PANIC when a=0 ∨ a>len(j)·2 ∨ a mod 2≠0 ∨ j_{(a/2)-1} ∉ basicblocks.
/// Checks: (1) a === HALT_ADDRESS → HALT, (2) a===0 || a>maxAddress || a%2!==0 → PANIC,
/// (3) index = (a/2)-1; index < 0 || index >= jumpTable.len() → PANIC, (4) validate_branch_target(target) → PANIC if invalid.
pub struct JumpIndInstruction;

impl JumpIndInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for JumpIndInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_JUMP_IND)
    }

    fn name(&self) -> &'static str {
        "JUMP_IND"
    }

    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_one_register_and_immediate(context.operands, context.fskip);
        let register_value = context.registers[parsed.register_a as usize];
        let a = (register_value.wrapping_add(parsed.immediate_x as u64)) & 0xffff_ffff;

        // Gray Paper: HALT when a = 2^32 - 2^16
        if a == u64::from(HALT_ADDRESS) {
            return InstructionResult::new(RESULT_CODE_HALT as i32, 0);
        }
        // Gray Paper: PANIC when a = 0 ∨ a > len(j)·2 ∨ a mod 2 ≠ 0
        let max_address = (context.jump_table.len() as u64) * 2;
        if a == 0 || a > max_address || (a % 2) != 0 {
            return InstructionResult::new(RESULT_CODE_PANIC as i32, 0);
        }
        // index = (a/2) - 1; PANIC if index < 0 (i.e. a/2 < 1) or index >= len
        let half = a / 2;
        if half < 1 {
            return InstructionResult::new(RESULT_CODE_PANIC as i32, 0);
        }
        let index = (half - 1) as usize;
        if index >= context.jump_table.len() {
            return InstructionResult::new(RESULT_CODE_PANIC as i32, 0);
        }
        let target = context.jump_table[index];
        if let Some(panic_result) = validate_branch_target(target, context.code, context.bitmask) {
            return panic_result;
        }
        context.program_counter = target;
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

impl Default for JumpIndInstruction {
    fn default() -> Self {
        Self::new()
    }
}

/// LOAD_IMM_JUMP (opcode 80): load immediate and jump. branch(immed_Y, ⊤), reg_A' = immed_X.
/// targetAddress = pc + signedOffset (immediateY sign-extended). Checks: validate_branch_target(target) → PANIC if invalid.
pub struct LoadImmJumpInstruction;

impl LoadImmJumpInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for LoadImmJumpInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_LOAD_IMM_JUMP)
    }

    fn name(&self) -> &'static str {
        "LOAD_IMM_JUMP"
    }

    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_register_and_two_immediates(context.operands, context.fskip);
        // Gray Paper: target = ι + signfunc(immed_Y); immediate_y is already sign-extended from parser
        let target = (context.program_counter as i64).wrapping_add(parsed.immediate_y) as u32;
        if let Some(panic_result) = validate_branch_target(target, context.code, context.bitmask) {
            return panic_result;
        }
        context.registers[parsed.register_a as usize] = parsed.immediate_x as u64;
        context.program_counter = target;
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

impl Default for LoadImmJumpInstruction {
    fn default() -> Self {
        Self::new()
    }
}

/// LOAD_IMM_JUMP_IND (opcode 180): djump((reg_B + immed_Y) mod 2^32), reg_A' = immed_X.
/// Order: read reg_B, set reg_A = immed_X, then djump. Same djump checks as JUMP_IND.
pub struct LoadImmJumpIndInstruction;

impl LoadImmJumpIndInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for LoadImmJumpIndInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_LOAD_IMM_JUMP_IND)
    }

    fn name(&self) -> &'static str {
        "LOAD_IMM_JUMP_IND"
    }

    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_two_registers_and_two_immediates(context.operands, context.fskip);
        // Gray Paper: read reg_B before overwriting reg_A
        let register_b_value = context.registers[parsed.register_b as usize];
        context.registers[parsed.register_a as usize] = parsed.immediate_x as u64;
        let a = (register_b_value.wrapping_add(parsed.immediate_y as u64)) & 0xffff_ffff;

        // Gray Paper: HALT when a = 2^32 - 2^16
        if a == u64::from(HALT_ADDRESS) {
            return InstructionResult::new(RESULT_CODE_HALT as i32, 0);
        }
        // Gray Paper: PANIC when a = 0 ∨ a > len(j)·2 ∨ a mod 2 ≠ 0
        let max_address = (context.jump_table.len() as u64) * 2;
        if a == 0 || a > max_address || (a % 2) != 0 {
            return InstructionResult::new(RESULT_CODE_PANIC as i32, 0);
        }
        // index = (a/2) - 1; PANIC if index < 0 or index >= len
        let half = a / 2;
        if half < 1 {
            return InstructionResult::new(RESULT_CODE_PANIC as i32, 0);
        }
        let index = (half - 1) as usize;
        if index >= context.jump_table.len() {
            return InstructionResult::new(RESULT_CODE_PANIC as i32, 0);
        }
        let target = context.jump_table[index];
        if let Some(panic_result) = validate_branch_target(target, context.code, context.bitmask) {
            return panic_result;
        }
        context.program_counter = target;
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

impl Default for LoadImmJumpIndInstruction {
    fn default() -> Self {
        Self::new()
    }
}
